### PR TITLE
[CGP-337] [StdLib] Removed duplicate names

### PR DESCRIPTION
**Pull request by effectfully**
This finally removes duplicate names from the stdlib in a very simple way: we just rename each term. Some terms already do not have duplicates, but we rename them as well, because it's too easy to make a mistake and decide that a term has no duplicates while it actually has them.

Holed data types can't be renamed that easily, but

1. they do not contain duplicate names as far as I see
2. we will remove `HoledType` anyway, so I don't bother dealing with them right now

Type checking of generated terms is turned on finally and everything type checks correctly.

We need a testing machinery that checks Plutus Core entities for not having duplicate names, but I'll add that later, here I just want to fix a critical bug.

I also removed `getBuiltinEnumCountNat`. It doesn't really belong to the stdlib. I think, I'll return it later in the form of an "interesting generator", but for now I'm just removing this noise.

### CODE REVIEW

File: language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs [hunk 1]
```diff
@@ -169,52 +160,6 @@ getBuiltinFoldList = do
...
--- > \(n k : nat) ->
--- >     foldrNat {nat -> list nat}
--- >         (\(rec : nat -> list nat) (n' : nat) -> cons {nat} n' (rec (succ n')))
--- >         (\(n' : nat) -> nil {nat})
--- >         k
--- >         n
---
--- @n'@ appears twice in the term, @nat@ appears several times in types,
--- so this is not really a definition with unique names.
-getBuiltinEnumCountNat :: Quote (Term TyName Name ())
```
Comment (michaelpj) [inline language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs]: why did you delete this?

Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs]: > I also removed `getBuiltinEnumCountNat`. It doesn't really belong to the stdlib. I think, I'll return it later in the form of an "interesting generator", but for now I'm just removing this noise.

Comment (michaelpj) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs]: :sweat_smile: I can read, I promise

Comment (effectfully) [reply language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs]: I thought that maybe you just missed that sentence (happens to me). That's not an stdlib function, it doesn't do anything useful. I planned to implement `enumFromTo` in terms of it, but then realized I better stop avoiding dealing with `integer`s. This function is a noise, so I removed it.

**Comment by effectfully** [comment]
> I do wish we could track this information on the type level somehow, so the stdlib could produce non-renamed terms and it would just be up to the consumer to rename them before the typechecker will accept them.

I usually like ideas like this, but not in this case. The thing is that whatever we return, it has to contain indices and run in the `Quote` monad. It's really just simpler to always return terms/types without duplicate names, because it does not complicate the API and the user doesn't have to think whether `Unique`s are unique or not. Unique `Unique`s are appropriate in all cases, even when we don't need them.

**Comment by effectfully** [comment]
If going to merge, please squash.
